src/proc/constants.rs

```diff
@@ -1,11 +1,12 @@
 use crate::{
     arena::{Arena, Handle},
-    ArraySize, Constant, ConstantInner, Expression, ScalarKind, ScalarValue, Type, UnaryOperator,
+    ArraySize, Constant, ConstantInner, Expression, ScalarKind, ScalarValue, Type, TypeInner,
+    UnaryOperator,
 };
 
 #[derive(Debug)]
 pub struct ConstantSolver<'a> {
-    pub types: &'a Arena<Type>,
+    pub types: &'a mut Arena<Type>,
```

Comment:
This is concerning. If the goal of this processor is to solve constants, it shouldn't mutate the types.

Reply:
I changed this in the last commit because matrix access produces vectors which is a composite and requires a type handle that might need to be created